src/MetadataManager/pathToHash.ts

```diff
@@ -0,0 +1,273 @@
+      const type: number = file[1];
+
+      if (type === 1) {
+        subPathToHash[name] = await PathToHash.generateHash(vscode.Uri.joinPath(uri, '/' + name));
+      } else if (type === 2 && name !== '.meta') {
+        subPathToHash[name] = await this.getSubPathToHash(vscode.Uri.joinPath(uri, '/' + name));
+      }
+    }
+
+    return subPathToHash;
+  }
+
+  public async validateMetadata(pathToHash: any) {
+    // 1. Changing pathToHash, a tree structure, into a one-dimensional structure
+    let flattenPathToHash = await this.getFlatPathToHash(pathToHash);
+
+    // 2. Create metadata if pathToHash exists but does not have actual metadata files,
+    // If pathToHash exists and there is a actual metadata file, but there is no path inside, create
+    // a path and data insde
+    await this.createMetadata(flattenPathToHash);
+
+    // 3. Replace is_deleted with true for all metadata not in pathToHash
+    await this.deleteMetadata(flattenPathToHash);
+  }
+
+  public async getFlatPathToHash(pathToHash: any) {
+    let flatPathTohash: any = {};
+
+    let queue = [];
+    for (let data in pathToHash) {
+      queue.push([data, pathToHash[data], data.toString()]);
+    }
+
+    while (queue.length !== 0) {
+      const subPathToHash = queue.pop();
+      if (subPathToHash === undefined) {
+        continue;
+      }
+      if (vscode.workspace.workspaceFolders === undefined) {
+        break;
+      }
+      let path =
+          vscode.Uri.joinPath(vscode.workspace.workspaceFolders[0].uri, subPathToHash[2]).path;
+      if (fs.lstatSync(path).isDirectory()) {
+        for (let key in subPathToHash[1]) {
+          queue.push([key, subPathToHash[1][key], subPathToHash[2] + '/' + key]);
+        }
+      } else {
+        flatPathTohash[subPathToHash[2]] = subPathToHash[1];
+      }
+    }
+    return flatPathTohash;
+  }
+
+  public async createMetadata(flattenPathToHash: any) {
+    for (let path in flattenPathToHash) {
+      const hash = flattenPathToHash[path];
+
+      let metadata: any = await Metadata.get(hash);
+
+      if (Object.keys(metadata).length === 0) {
+        await Metadata.set(hash, {});
+
+      } else if (metadata[path]) {
+        return;
+      }
+
+      const filename: any = path.split('/').pop();
+      if (vscode.workspace.workspaceFolders === undefined) {
+        return;
+      }
+      const stats: any = await Metadata.getStats(
+          vscode.Uri.joinPath(vscode.workspace.workspaceFolders[0].uri, path));
+      metadata[path] = {};
+      metadata[path]['name'] = filename;
+      metadata[path]['file-extension'] = filename.split('.')[1];
+      metadata[path]['create-time'] = stats.birthtime;
+      metadata[path]['modified-time'] = stats.mtime;
+      metadata[path]['is-deleted'] = false;
+      await Metadata.set(hash, metadata);
+    }
+  }
+
+  public async deleteMetadata(flattenpathToHash: any) {
+    if (vscode.workspace.workspaceFolders === undefined) {
+      return;
+    }
```

Comment:
It seems that PathToHash class will be called by MetadataManager, if so, let's not assume that we have no workspaceFolders. Instead, use 'assert'.